### PR TITLE
Add importer for Open SLR 88 a.k.a Att-HACK

### PR DESCRIPTION
**Pull request by wasertech**
# Open SLR 88: Att-HACK

> This data is acted expressive speech in French, 100 phrases with multiple versions (3 to 5) in four social attitudes : friendly, distant, dominant and seductive.

> This research has been supported by the French Ph2D/IDF MoVE project on MOdelling of speech attitudes and application to an expressive conversationnal agent and funded by the Ile- De-France region. 

```
@inproceedings{lemoine:hal-02508362,
 TITLE = {{Att-HACK: An Expressive Speech Database with Social Attitudes}},
 AUTHOR = {Le Moine, Cl{\'e}ment and Obin, Nicolas},
 URL = {https://hal.archives-ouvertes.fr/hal-02508362},
 BOOKTITLE = {{Speech Prosody}},
 ADDRESS = {Tokyo, Japan},
 YEAR = {2020},
 MONTH = May,
 KEYWORDS = {expressivity ; speech prosody ; social attitude ; speech database ; Index Terms: expressivity},
 PDF = {https://hal.archives-ouvertes.fr/hal-02508362/file/Att_HACK_A_Speech_Database_for_the_Study_of_Social_Attitudes_16_03_20.pdf},
 HAL_ID = {hal-02508362},
 HAL_VERSION = {v1},
}
```

```
================
== TensorFlow ==
================

NVIDIA Release 20.06-tf1 (build 13409399)
TensorFlow Version 1.15.2

Container image Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
Copyright 2017-2020 The TensorFlow Authors.  All rights reserved.

NVIDIA Deep Learning Profiler (dlprof) Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.

Various files include modifications (c) NVIDIA CORPORATION.  All rights reserved.
NVIDIA modifications are covered by the license terms that apply to the underlying project or file.
WARNING: Detected NVIDIA NVIDIA GeForce RTX 3060 GPU, which is not yet supported in this version of the container

NOTE: MOFED driver for multi-node communication was not detected.
      Multi-node communication performance may be reduced.

root@e22f7c6151ac:/code# python bin/import_slr88.py /mnt/extracted/data/Att-HACK/
2021-12-31 02:12:58.579105: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.11.0
WARNING:tensorflow:Deprecation warnings have been disabled. Set TF_ENABLE_DEPRECATION_WARNINGS=1 to re-enable them.
WARNING: No --validate_label_locale specified, your might end with inconsistent dataset.
Found archive "/mnt/extracted/data/Att-HACK/txt.tgz" - not downloading.
Found archive "/mnt/extracted/data/Att-HACK/wav.tgz" - not downloading.
Found directory "/mnt/extracted/data/Att-HACK/Volumes/CLEM_HDD/IRCAM/Open_SLR/txt" - not extracting it from archive.
Found directory "/mnt/extracted/data/Att-HACK/Volumes/CLEM_HDD/IRCAM/Open_SLR/wav" - not extracting it from archive.
Importing WAV files...
Progress |#################################################################################################### |  99% completedImported 36143 samples.
Skipped 5 samples that were too short to match the transcript.
Skipped 486 samples that were longer than 15 seconds.
Final amount of imported audio: 72:53:18 from 75:13:55.
```

Which produces:

```
root@e22f7c6151ac:/code# ls -la /mnt/extracted/data/Att-HACK/*.csv
-rw-r--r-- 1 root root  415536 Dec 31 02:14 /mnt/extracted/data/Att-HACK/Att-HACK_dev.csv
-rw-r--r-- 1 root root  415484 Dec 31 02:14 /mnt/extracted/data/Att-HACK/Att-HACK_test.csv
-rw-r--r-- 1 root root 3320994 Dec 31 02:14 /mnt/extracted/data/Att-HACK/Att-HACK_train.csv
root@e22f7c6151ac:/code# cat /mnt/extracted/data/Att-HACK/Att-HACK_train.csv | head
wav_filename,wav_filesize,transcript
/mnt/extracted/data/Att-HACK/Volumes/CLEM_HDD/IRCAM/Open_SLR/wav/M19_a2_s047_v07.wav,271700,nous avons eu peur
/mnt/extracted/data/Att-HACK/Volumes/CLEM_HDD/IRCAM/Open_SLR/wav/M17_a4_s100_v05.wav,413026,c est vrai allons prendre un café
/mnt/extracted/data/Att-HACK/Volumes/CLEM_HDD/IRCAM/Open_SLR/wav/M12_a3_s023_v02.wav,248768,évidemment
/mnt/extracted/data/Att-HACK/Volumes/CLEM_HDD/IRCAM/Open_SLR/wav/F18_a3_s014_v03.wav,263084,impossible
/mnt/extracted/data/Att-HACK/Volumes/CLEM_HDD/IRCAM/Open_SLR/wav/F06_a1_s095_v04.wav,229704,vous avez dormi toute la nuit
/mnt/extracted/data/Att-HACK/Volumes/CLEM_HDD/IRCAM/Open_SLR/wav/M13_a1_s096_v04.wav,448100,vous avez dit ce qu'il fallait
/mnt/extracted/data/Att-HACK/Volumes/CLEM_HDD/IRCAM/Open_SLR/wav/F18_a3_s055_v03.wav,339884,tu as pris la fuite
/mnt/extracted/data/Att-HACK/Volumes/CLEM_HDD/IRCAM/Open_SLR/wav/M13_a3_s008_v01.wav,190556,tant pis
/mnt/extracted/data/Att-HACK/Volumes/CLEM_HDD/IRCAM/Open_SLR/wav/M04_a3_s045_v02.wav,253642,je suis parti tôt
/mnt/extracted/data/Att-HACK/Volumes/CLEM_HDD/IRCAM/Open_SLR/wav/M19_a2_s054_v01.wav,527480,faisons leur confiance
/mnt/extracted/data/Att-HACK/Volumes/CLEM_HDD/IRCAM/Open_SLR/wav/M13_a2_s055_v05.wav,384596,tu as pris la fuite
/mnt/extracted/data/Att-HACK/Volumes/CLEM_HDD/IRCAM/Open_SLR/wav/F08_a1_s006_v04.wav,172878,bonjour
/mnt/extracted/data/Att-HACK/Volumes/CLEM_HDD/IRCAM/Open_SLR/wav/F02_a3_s033_v03.wav,335204,je pars demain
/mnt/extracted/data/Att-HACK/Volumes/CLEM_HDD/IRCAM/Open_SLR/wav/F01_a4_s040_v04.wav,379220,donc fais moi signe
```

### CODE REVIEW

**Comment by wasertech** [comment]
@lissyx What do think about it?

**Comment by wasertech** [comment]
Should we keep audio file longer than 15s?

**Comment by lissyx** [comment]
> Should we keep audio file longer than 15s?

no, it can have a detrimental effect on training depending on GPU memory size and batching. I would suggest we do as someplace elses, keep a parameter that controls maximum length actionable from CLI so people who have verified that it can work, can leverage this amount of data, but we have a good default behavior for whatever is considered as "low bar of support".

**Review by lissyx — APPROVED** [review]
It looks sane to me, but I might have not followed closely the changes to the codebase, and some assumptions might not hold anymore. @reuben should verify as well.

File: bin/import_slr88.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+        with open(
+            target_csv_template.format("train"), "w", encoding="utf-8", newline=""
+        ) as train_csv_file:  # 80%
+            with open(
+                target_csv_template.format("dev"), "w", encoding="utf-8", newline=""
+            ) as dev_csv_file:  # 10%
+                with open(
+                    target_csv_template.format("test"), "w", encoding="utf-8", newline=""
+                ) as test_csv_file:  # 10%
```
Comment (reuben) [inline bin/import_slr88.py:156]: nit: you can open multiple context managers with comma:
```suggestion
        with open(
            target_csv_template.format("train"), "w", encoding="utf-8", newline=""  # 80%
        ) as train_csv_file, open(
            target_csv_template.format("dev"), "w", encoding="utf-8", newline=""  # 10%
        ) as dev_csv_file, open(
            target_csv_template.format("test"), "w", encoding="utf-8", newline=""  # 10%
        ) as test_csv_file:
```

File: bin/import_slr88.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+                        if not transcript:
+                            continue
+                        wav_filename = item[0]
+                        i_mod = i % 10
+                        if i_mod == 0:
+                            writer = test_writer
+                        elif i_mod == 1:
+                            writer = dev_writer
+                        else:
+                            writer = train_writer
```
Comment (reuben) [inline bin/import_slr88.py:175]: 80%/10%/10% split usually leads to large validation and test sets. What we prefer to do in our datasets is a sample size power calculation. See for example `get_sample_size` in `bin/import_fisher.py`.

**Review by reuben — COMMENTED** [review]
Other than what @lissyx already said, looks good. I only have a couple of comments. Thanks for the PR!

File: bin/import_slr88.py (hunk 1 above)
Comment (wasertech) [reply bin/import_slr88.py]: Yes much tidier indeed. Thanks for the tip.

File: bin/import_slr88.py (hunk 2 above)
Comment (wasertech) [reply bin/import_slr88.py]: [cd41cc9](https://github.com/coqui-ai/STT/pull/2062/commits/cd41cc9e97fe348d36defe2e81883ff9b1cfc693): It's an half assed job but it should give us the distribution we are looking for.

**Comment by wasertech** [comment]
So I tried one epoch with SLR88 to test and was greeted with [this lovely error](https://gist.github.com/wasertech/0004c6e6aed855b2e42a4ca5c2ba0116) during validation.

I still need to fine-tune the importer `file size / transcription length` ratio.

**Comment by wasertech** [comment]
The dataset uses Microsoft Extended WAV Format. I used `ffmpeg` to convert the files next the original ones. (`{filename}_.wav`)
I got to save a best_dev model so it looks right.

File: bin/import_slr88.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
+#!/usr/bin/env python3
+import csv
+import os
+import random
+import subprocess
+import tarfile
+import unicodedata
+import sox
```
Comment (wasertech) [inline bin/import_slr88.py:8]: I think we could remove this dependency then.

File: bin/import_slr88.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+from glob import glob
+from multiprocessing import Pool
+
+import progressbar
```
Comment (wasertech) [inline bin/import_slr88.py:12]: We could use `tqdm`.

File: bin/import_slr88.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+        else:
+            raise FileNotFoundError
+    except FileNotFoundError:
+        print('No directory "%s" - extracting archive...' % extracted_path)
+        tar = tarfile.open(archive_path)
+        tar.extractall(target_dir)
+        tar.close()
+
+def one_sample(sample):
+    """ Take a audio file, and optionally convert it to 16kHz mono channel WAV """
```
Comment (wasertech) [inline bin/import_slr88.py:66]: an audio file

**Comment by reuben** [comment]
Feel free to re-request review once you've addressed your notes above. Also, I'd rather name the file `import_atthack.py` since that's the actual name of the dataset.

**Comment by wasertech** [comment]
> Feel free to re-request review once you've addressed your notes above. Also, I'd rather name the file `import_atthack.py` since that's the actual name of the dataset.

For consistency we should also change import_slr57.py (African French) then.

I’ll let you know once I’m happy with it.

Thanks 👍🏻

File: bin/import_slr88.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+        counter["invalid_label"] += 1
+    elif int(frames / SAMPLE_RATE * 1000 / 15 / 2) < len(str(label)):
+        # Excluding samples that are too short to fit the transcript
+        counter["too_short"] += 1
+    elif frames / SAMPLE_RATE > MAX_SECS:
+        # Excluding very long samples to keep a reasonable batch-size
+        counter["too_long"] += 1
+    else:
+        # This one is good - keep it for the target CSV
+        rows.append((wav_filename, file_size, label))
```
Comment (wasertech) [inline bin/import_slr88.py:98]: I would prefer to produce `rows.append((wav_filename.relative_to(os.path.join(target_dir, ARCHIVE_ROOT_PATH)), file_size, label))` to get relative path in the csv files.

**Comment by reuben** [comment]
Merged, thanks for the PR!
